### ISSUE TITLE
Bug in Zonemaster::Test::Syntax (syntax04, syntax07 and syntax08)

### ISSUE THREAD

**Issue by pawal** [dotse/zonemaster-engine#12]
_From @vlevigneron on October 14, 2014 15:28_

With very broken testing zones zone02.zut-root.rd.nic.fr and zone01.zut-root.rd.nic.fr (look at SOA MNAME, MX AND NS list) I discovered that the '.' case was not properly handled... I'll fix it today and will add proper error messages for that new cases...

_Copied from original issue: dotse/zonemaster#107_

**Comment by pawal** [comment]
_From @vlevigneron on October 14, 2014 17:6_

The following error codes hve been added and here is the proposed severity level.

      "MNAME_IS_ROOT" : "ERROR",
      "MNAME_NON_ALLOWED_CHARS" : "ERROR",
      "NAMESERVER_IS_ROOT" : "ERROR",
      "NAMESERVER_NON_ALLOWED_CHARS" : "ERROR",
      "MX_IS_ROOT" : "ERROR",
      "MX_NON_ALLOWED_CHARS" : "ERROR",

We already have the following one in the current code:

      "NON_ALLOWED_CHARS" : "ERROR",

**Comment by pawal** [comment]
_From @einar on October 15, 2014 7:30_

Vincent please take into consideration the documented error levels here:

"https://github.com/dotse/zonemaster/blob/master/docs/specifications/tests/SeverityLevelDefinitions.md". I really do not see how you could have these levels on these messages as they will most certainly "go unnoticed by the user" and probably has zero impact on the function of the zone at all (especially MX).

I was just planning on updating the policy to more reasonable values (after the discussion on our last meeting) but these values above need to be checked too. Can we please default to a LOW policy instead of claiming all new tests to be at a high level? It's much easier to raise than to lower and all of these can never be more than a WARNING according to Calles document (I think, some may be ERROR)...

I will update them but please please take more consideration into policy changes in the future. With these values above we'd have an extra 10-20% ERRORS in the .SE-zone without reason, for starters. ;P

**Comment by pawal** [comment]
_From @cdybedahl on October 15, 2014 7:35_

Why does there need to be special cases for root at all? Is this something that Zonecheck does?

**Comment by pawal** [comment]
_From @cdybedahl on October 15, 2014 7:57_

There is no support for the *_IS_ROOT messages in either the requirements or specifications.

**Comment by pawal** [comment]
_From @vlevigneron on October 15, 2014 8:49_

I agree that all return values should have been specified in test
specifications to have common discussions on each of them.

le 15 oct., Calle Dybedahl a ?crit :
> There is no support for the *_IS_ROOT messages in either the requirements or specifications.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dotse/zonemaster/issues/107#issuecomment-59170174

-- 
	Vincent Levigneron  A.F.N.I.C.  Vincent.Levigneron@nic.fr

**Comment by pawal** [comment]
_From @vlevigneron on October 15, 2014 8:52_

I guess this is a wording issue, the roor is not the issue but do we
consider that that kind of zone should not raise a warning oe notice
message  ?

> dig +multi zone02.zut-root.rd.nic.fr mx

; <<>> DiG 9.9.0-RedHat-9.9.0-4.fc17 <<>> +multi
zone02.zut-root.rd.nic.fr mx
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 26002
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 3, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;zone02.zut-root.rd.nic.fr. IN MX

;; ANSWER SECTION:
zone02.zut-root.rd.nic.fr. 1800	IN MX 10 .

;; AUTHORITY SECTION:
zone02.zut-root.rd.nic.fr. 1800	IN NS .
zone02.zut-root.rd.nic.fr. 1800	IN NS ns334987.ip-46-105-116.eu.
zone02.zut-root.rd.nic.fr. 1800	IN NS ns324830.ip-178-33-232.eu.

;; Query time: 2313 msec
;; SERVER: 212.27.40.240#53(212.27.40.240)
;; WHEN: Wed Oct 15 10:47:05 2014
;; MSG SIZE  rcvd: 158

le 15 oct., Calle Dybedahl a ?crit :
> Why does there need to be special cases for root at all? Is this something that Zonecheck does?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dotse/zonemaster/issues/107#issuecomment-59168311

-- 
	Vincent Levigneron  A.F.N.I.C.  Vincent.Levigneron@nic.fr

**Comment by pawal** [comment]
_From @vlevigneron on October 15, 2014 9:10_

le 15 oct., Einar Lönn a ?crit :
> Vincent please take into consideration the documented error levels here:
> 
> "https://github.com/dotse/zonemaster/blob/master/docs/specifications/tests/SeverityLevelDefinitions.md". I really do not see how you could have these levels on these messages as they will most certainly "go unnoticed by the user" and probably has zero impact on the function of the zone at all (especially MX).
> 

That's exactly why I asked a review on these new error codes. I chose
the default in compliance with "NON_ALLOWED_CHARS" which share some
similarities with these errors. 

> I was just planning on updating the policy to more reasonable values (after the discussion on our last meeting) but these values above need to be checked too. Can we please default to a LOW policy instead of claiming all new tests to be at a high level? It's much easier to raise than to lower and all of these can never be more than a WARNING according to Calles document (I think, some may be ERROR)...
> 

Are you sure you would have noticed these new codes I I had chosen a
"INFO" level ?

By the end of this month we should compare policy version we had during
our face to face meeting and the last version we plan to release. There
will be some new codes following new test cases implemented and all
DNSSEC part was missing. For all codes reviewed during our meeting, the
severity level should have been already fixed in current version. But As
eveybody can modify the code and policy, I highly recommend that we all
check if it's conform with what was decided.


> I will update them but please please take more consideration into policy changes in the future. With these values above we'd have an extra 10-20% ERRORS in the .SE-zone without reason, for starters. ;P
> 

I guess we should add the following requirement:

"Tool results should not add more than an extra +2% ERRORS in .fr and .se
zones." :-)

> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dotse/zonemaster/issues/107#issuecomment-59167798

-- 
	Vincent Levigneron  A.F.N.I.C.  Vincent.Levigneron@nic.fr

**Comment by pawal** [comment]
_From @cdybedahl on October 15, 2014 9:10_

A zone with . listed as an NS should fail nameserver06 (name cannot be resolved). For MX, we've chosen to only test that such a record exists and does not point at a CNAME. Adding a test that the MX actually resolves to an address would make sense, perhaps. Special-casing for the root zone does not make sense, since the problem isn't that it's the root zone. The problem is that a useful IP address cannot be found via the name.

Differently put, why should we report a problem for the root zone but not for .COM or .XN--VERMGENSBERATUNG-PWB (to pick two names more or less at random)?

**Comment by pawal** [comment]
_From @cdybedahl on October 15, 2014 9:17_

> Are you sure you would have noticed these new codes I I had chosen a "INFO" level ?

No, because you put their addition in a couple of commits totaling almost 2000 lines of changes and with the commit messages "Zonemaster::Test::Connectivity french translation, Unit tests" and "Zonemaster::Test::Syntax french translation". Neither of those make it at all obvious that entirely new messages were being added and new tests suggested, and the sheer mass of changes made it hard to spot manually.

**Comment by pawal** [comment]
_From @vlevigneron on October 15, 2014 9:18_

That's what I said, it's just a wording/level issue. If I change, "_IS_ROOT", by "_IS_DOT" or even something completely different, no matter, do you feel more comfortable ? I guess we should at least notice the user of these specific cases ?

**Comment by pawal** [comment]
_From @vlevigneron on October 15, 2014 9:27_

That's why I opened this issue with the list of new return codes with
there proposed severity level...
*
le 15 oct., Calle Dybedahl a ?crit :
> > Are you sure you would have noticed these new codes I I had chosen a "INFO" level ?
> 
> No, because you put their addition in a couple of commits totaling almost 2000 lines of changes and with the commit messages "Zonemaster::Test::Connectivity french translation, Unit tests" and "Zonemaster::Test::Syntax french translation". Neither of those make it at all obvious that entirely new messages were being added and new tests suggested, and the sheer mass of changes made it hard to spot manually.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dotse/zonemaster/issues/107#issuecomment-59178566

-- 
	Vincent Levigneron  A.F.N.I.C.  Vincent.Levigneron@nic.fr

**Comment by pawal** [comment]
_From @cdybedahl on November 3, 2014 14:41_

This issue is, as I understand it, blocked on a discussion about if we should add a special test to see if various RDATA fields holding domain names are set to the root zone. As such, I don't think it should block the alpha release (and probably be postponed to after 1.0).

**Comment by pawal** [comment]
I changed this to Post-1.0 pending any clarifying discussion on the issue.

**Comment by pawal** [comment]
_From @vlevigneron on November 3, 2014 14:51_

I agree it can be postponed. While results when this kind of case happens are rather
"confusing", this is not a bug and it can be seen as an improvement to clarify messages in a future
version.

le 03 nov., Calle Dybedahl a ?crit :
> This issue is, as I understand it, blocked on a discussion about if we should add a special test to see if various RDATA fields holding domain names are set to the root zone. As such, I don't think it should block the alpha release (and probably be postponed to after 1.0).
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dotse/zonemaster/issues/107#issuecomment-61486683

-- 
	Vincent Levigneron  A.F.N.I.C.  Vincent.Levigneron@nic.fr

**Comment by sandoche2k** [comment]
@vlevigneron - even though this task has been assigned for .3, do you think it will be possible for you to complete by .2 (15/06/2015)?
If yes - resource required and prospective date.
If no - we will keep it for release .3

**Comment by sandoche2k** [comment]
@vlevigneron - as mentioned during the last conf. call please comment on time required to solve this issue.

**Comment by pawal** [comment]
Closing. @sandoche2k creates a new issue on "do dot or not" (labels are sometimes logged with suffixed dot, sometimes not).
